### ISSUE TITLE
Use libimagequant for 8-bit PNG generation

### ISSUE THREAD

**Issue by pornel** [mapnik/mapnik#2706]
There are several issues related to low quality of PNG8 palettes: #2292 #2497 #2110

I'd like to recommend you to have a look at [libimagequant](http://pngquant.org/lib/), which is an embeddable C library for generation of *high-quality* palettes. The API allows deep integration, so you generate palette in memory without having to create any temporary PNG files.

Here's an example: 585KB down to 264KB:

![map-png24](https://cloud.githubusercontent.com/assets/72159/6204201/2570766e-b53a-11e4-9453-dece37cbc73f.png)

![map-png8](https://cloud.githubusercontent.com/assets/72159/6204202/2571c23a-b53a-11e4-9d70-b9680d0a5e90.png)

The output PNG is less than a half size and doesn't [suffer from banding, unlike palettes generated with old algorithms](http://pngquant.org/vsphotoshop.html) that always throw away least significant bits.

**Comment by springmeyer** [comment]
@pornel - thank you very much for the suggestion: `libimagequant` looks impressive. If the speed is as good as Mapnik's internal code (octree.hpp and hextree.hpp) then I think moving to `libimagequant` would be possible. The lack of banding would be a great improvement.

**Comment by pornel** [comment]
On x86 with SSE the speed is comparable with other quantizers, and if it's not fast enough you can use `liq_set_speed()` to disable some expensive minor enhancements.

**Comment by springmeyer** [comment]
@pornel - thanks. I'm concerned that it appears libimagequant does not have unit tests. Is this the case?

**Comment by pornel** [comment]
Yes. I have a large functional test suite which tracks regressions in quality, but can't redistribute it due to images being copyrighted.

**Comment by der-stefan** [comment]
Yes, I would really like to have this feature. Currently, our map server saves 8-bit tiles, since 32-bit consumes too much disk space. But we get ugly color rings. See #2497 and openstreetmap/mod_tile/issues/64

**Comment by springmeyer** [comment]
Okay, next action here is for someone to create an experimental branch of mapnik that tries integrating `libimagequant` and assess performance. One easy way is to use the `image_encoding_speed_test.py` like https://github.com/mapnik/mapnik/pull/2671#issuecomment-72548546. Before for reference this lives over in the python bindings now: https://github.com/mapnik/python-mapnik/blob/master/test/python_tests/image_encoding_speed_test.py

**Comment by danpat** [comment]
I bolted on `libimagequant` and added two new parameters `m=iq` and `iqs=N` to the PNG encoder.

The quick summary is that the encoded images look better, are slightly bigger (understandable with the extra dithering), and quantization is slower, how much depends on the `iqs` setting.  Looking at the numbers below, I doubt `libimagequant` can compete with `hextree` or `octtree` on pure speed.  There are a couple of inefficiencies in my implementation, but not enough to make up the difference (one small bottleneck is that `libimagequant` expects RGBA byte order, but `mapnik::image` is storing ABGR).

The upside is that `libimagequant` has no external dependencies and is easy enough to integrate.  It could be included as an option for those that prefer nice dithering to outright performance.  It could be that file size improvements would be more obvious on larger tiles, but I haven't tried anything other than `image_encoding_speed_test.py` at this stage.

Here are the results comparing `hextree`, `octtree` and `libimagequant` using `image_encoding_speed_test.py`:

    min: 1.4510ms | avg: 1.9840ms | total: 20.075ms | len: 139 <-- blank png8:m=h
    min: 2.0911ms | avg: 2.3404ms | total: 23.602ms | len: 139 <-- blank png8:m=o
    min: 10.949ms | avg: 12.558ms | total: 125.98ms | len: 362 <-- blank png8:m=iq:iqs=10
    min: 23.355ms | avg: 25.801ms | total: 258.40ms | len: 362 <-- blank png8:m=iq:iqs=5
    min: 23.376ms | avg: 27.306ms | total: 273.50ms | len: 362 <-- blank png8:m=iq

    min: 11.461ms | avg: 12.890ms | total: 129.08ms | len: 126 <-- solid png8:m=o
    min: 12.042ms | avg: 12.337ms | total: 123.59ms | len: 126 <-- solid png8:m=h
    min: 12.381ms | avg: 13.858ms | total: 138.95ms | len: 349 <-- solid png8:m=iq:iqs=10
    min: 23.852ms | avg: 27.861ms | total: 279.03ms | len: 349 <-- solid png8:m=iq
    min: 27.985ms | avg: 31.307ms | total: 313.55ms | len: 349 <-- solid png8:m=iq:iqs=5

    min: 8.7101ms | avg: 9.5679ms | total: 95.926ms | len: 33731 <-- many_colors png8:m=h
    min: 8.7289ms | avg: 9.8906ms | total: 99.195ms | len: 33728 <-- many_colors png8:m=o
    min: 37.878ms | avg: 41.733ms | total: 417.74ms | len: 33647 <-- many_colors png8:m=iq:iqs=10
    min: 41.641ms | avg: 47.273ms | total: 473.18ms | len: 33722 <-- many_colors png8:m=iq:iqs=5
    min: 41.900ms | avg: 48.368ms | total: 484.12ms | len: 33722 <-- many_colors png8:m=iq

    min: 9.7370ms | avg: 12.049ms | total: 120.86ms | len: 43267 <-- aerial_24 png8:m=o
    min: 19.484ms | avg: 22.740ms | total: 227.93ms | len: 46506 <-- aerial_24 png8:m=h
    min: 87.608ms | avg: 93.404ms | total: 934.42ms | len: 52509 <-- aerial_24 png8:m=iq:iqs=10
    min: 588.20ms | avg: 610.21ms | total: 6102.6ms | len: 52087 <-- aerial_24 png8:m=iq:iqs=5
    min: 1014.4ms | avg: 1038.0ms | total: 10381.ms | len: 51908 <-- aerial_24 png8:m=iq

The aerial_24 test image shows the improved dithering the best:

original | `hextree` | `libimagequant` speed `3` (default) | `libimagequant` speed `10` (fastest)
----- | ------ | ------ | ------
![12_654_1580](https://cloud.githubusercontent.com/assets/1892250/7626891/f7f61fb6-f9c4-11e4-802b-d618e5738cf5.png) | ![aerial_24_png8 m h](https://cloud.githubusercontent.com/assets/1892250/7626812/df146148-f9c3-11e4-974b-cfa54ae03826.png) | ![aerial_24_png8 m iq](https://cloud.githubusercontent.com/assets/1892250/7626815/e91c3008-f9c3-11e4-9411-488f8bb059ec.png) | ![aerial_24_png8 m iq iq 10](https://cloud.githubusercontent.com/assets/1892250/7626813/e674874c-f9c3-11e4-9ef7-082c37720b64.png)
 | Note posterization on ocean | |

**Comment by springmeyer** [comment]
@danpat - incredible work putting this together. Yes, I think it's worth continuing to polish the patch to include given the improved quality and the ease with which you've shown we can integrate libimagequant as an optional encoder. I wish the `iqs=10` speeds were better but that is a lot to ask given the `hextree` encoder has seen a lot of optimization and profiling effort.

Ideas for next steps:

  - There is a large image over at https://github.com/Automattic/node-canvas/issues/306 which is an interesting stress test of PNG. It's an image that makes more sense encoded as JPEG or Webp, but I'm curious if the relative perf of hextree/octree vs libimagequant is these same as with the `aerial_24` or any different.
  - Given a testcase where libimagequant is slower, I'm interested what profiling shows libimagequant doing compared to hextree. Maybe differences in where time is taken could help us figure out some optimizations. Note: I find profiling via a python script tricky on OS X so I tend to write pure C++ benchmarks. There is one over at https://github.com/mapnik/mapnik/blob/master/benchmark/test_png_encoding2.cpp you can co-opt. Uncomment [this line](https://github.com/mapnik/mapnik/blob/master/benchmark/build.py#L31) and do `./configure BENCHMARK=True` to trigger the creation of `./benchmark/out/test_png_encoding2`.

In terms of the todo's you listed at:  https://github.com/danpat/mapnik/commit/01380588ba5f5e3adc4e5e09609e827c40297773: 

> 1 & 4 bit encodings
   
:+1: 

> proper Mason integration

I think libimagequant is small enough to bundle into deps/libimagequant at least for now.

**Comment by springmeyer** [comment]
Small note: we are close to feature freeze on upcoming 3.0.0 release. Had this against 3.0.0 so it would not get lost. This would be a great new feature in an upcoming tag, so for now un-tagging the 3.0.0 milestone.

**Comment by danpat** [comment]
Here's the results of running `libimagequant` against the image at https://github.com/Automattic/node-canvas/issues/306

    min: 3389.7ms | avg: 3562.5ms | total: 35626.ms | len: 12046558 <-- burning_man png8:m=o
    min: 3452.9ms | avg: 3598.6ms | total: 35986.ms | len: 11933658 <-- burning_man png8:m=h
    min: 24447.ms | avg: 24944.ms | total: 249441ms | len: 15440900 <-- burning_man png8:m=iq:iqs=10
    min: 29169.ms | avg: 29512.ms | total: 295130ms | len: 15294251 <-- burning_man png8:m=iq:iqs=5
    min: 32941.ms | avg: 33581.ms | total: 335818ms | len: 15264144 <-- burning_man png8:m=iq

So overall, no competition to `hextree` for speed.  The output palettes look quite different, `iqs=10` has lost some of the small but bright colors (red tent tops) so it looks a little washed out:

 `hextree` | `m=iq` | `m=iq:iqs=10`
---- | ---- | ----
![screen shot 2015-05-14 at 4 08 06 pm](https://cloud.githubusercontent.com/assets/1892250/7643981/784ba936-fa53-11e4-8207-1bfd623fd303.png) | ![screen shot 2015-05-14 at 4 23 15 pm](https://cloud.githubusercontent.com/assets/1892250/7644142/adaabee4-fa55-11e4-80e2-9aa0e99bf83f.png) | ![screen shot 2015-05-14 at 4 08 30 pm](https://cloud.githubusercontent.com/assets/1892250/7643986/84d664ca-fa53-11e4-9640-28aee499bda7.png)

**Comment by danpat** [comment]
Some good news on the performance front.  I added `libimagequant` to `deps` and integrated it with the `scons` build.  In the process, I realized that the makefile it came with wasn't detecting SSE on Darwin, so the numbers above are not using CPU optimizations.  Below are the results when built with SSE, it's much more competitive with `hextree`, although still not quite there:

    min: 1.4481ms | avg: 2.1733ms | total: 22.171ms | len: 139 <-- blank png8:m=h
    min: 6.8049ms | avg: 9.0181ms | total: 90.783ms | len: 362 <-- blank png8:m=iq:iqs=10
    min: 18.002ms | avg: 21.031ms | total: 210.79ms | len: 362 <-- blank png8:m=iq
    min: 19.885ms | avg: 22.577ms | total: 226.28ms | len: 362 <-- blank png8:m=iq:iqs=5

    min: 8.6309ms | avg: 10.111ms | total: 101.71ms | len: 349 <-- solid png8:m=iq:iqs=10
    min: 12.761ms | avg: 14.272ms | total: 143.21ms | len: 126 <-- solid png8:m=h
    min: 19.729ms | avg: 22.221ms | total: 222.82ms | len: 349 <-- solid png8:m=iq
    min: 21.203ms | avg: 23.042ms | total: 231.01ms | len: 349 <-- solid png8:m=iq:iqs=5

    min: 10.435ms | avg: 11.825ms | total: 118.77ms | len: 33731 <-- many_colors png8:m=h
    min: 15.459ms | avg: 19.691ms | total: 197.50ms | len: 33647 <-- many_colors png8:m=iq:iqs=10
    min: 20.637ms | avg: 22.552ms | total: 226.16ms | len: 33722 <-- many_colors png8:m=iq:iqs=5
    min: 23.869ms | avg: 25.198ms | total: 252.74ms | len: 33722 <-- many_colors png8:m=iq

    min: 20.866ms | avg: 27.232ms | total: 272.89ms | len: 46506 <-- aerial_24 png8:m=h
    min: 32.623ms | avg: 37.005ms | total: 370.66ms | len: 52509 <-- aerial_24 png8:m=iq:iqs=10
    min: 205.36ms | avg: 220.46ms | total: 2205.0ms | len: 52087 <-- aerial_24 png8:m=iq:iqs=5
    min: 348.84ms | avg: 374.46ms | total: 3745.3ms | len: 51908 <-- aerial_24 png8:m=iq

I've added some SSE detection logic and a flag to SConstruct.  If built on a platform that supports SSE (currently only detecting Linux and OS X) with the default flag, it will compile with SSE optimizations enabled.  Cross-compiling will have them disabled by default, but can be overridden.

**Comment by pornel** [comment]
In https://github.com/pornel/pngquant/commit/a8598d9cc62414c53543e94ac7f404a01e1bcce3 I've implemented a new color search algorithm which is 20% faster with SSE and 2x faster without SSE.

**Comment by springmeyer** [comment]
@pornel - great, thanks for the heads up. We'll take a look at integrating and running the benchmarks again.

**Comment by der-stefan** [comment]
Is this new feature used in Mapnik 3.0?

**Comment by flippmoke** [comment]
@der-stefan no, not yet

**Comment by der-stefan** [comment]
Sorry for bothering - any news here?
